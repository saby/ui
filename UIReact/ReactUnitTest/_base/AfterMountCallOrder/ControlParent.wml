<ReactUnitTest._base.AfterMountCallOrder.ControlChild1/>
