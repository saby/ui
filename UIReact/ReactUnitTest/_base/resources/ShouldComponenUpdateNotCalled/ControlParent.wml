<ReactUnitTest._base.resources.ShouldComponenUpdateNotCalled.ControlChild testOption="{{ _testOption }}" />
