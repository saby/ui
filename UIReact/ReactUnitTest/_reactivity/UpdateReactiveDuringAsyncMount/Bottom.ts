import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_reactivity/UpdateReactiveDuringAsyncMount/Bottom';

export default class Bottom extends Control {
    _template: TemplateFunction = template;

    protected _beforeMount(): Promise<void> | void {
        return new Promise((resolve) => {
            setTimeout(() => {
                resolve();
            }, 20);
        });
    }
}
