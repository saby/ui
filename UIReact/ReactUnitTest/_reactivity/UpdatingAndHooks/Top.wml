<ReactUnitTest._reactivity.UpdatingAndHooks.Middle topValue="{{topValue}}" order="{{_options.order}}"/>
