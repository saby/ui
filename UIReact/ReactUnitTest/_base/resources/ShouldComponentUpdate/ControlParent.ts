import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_base/resources/ShouldComponentUpdate/ControlParent';
import ControlChild from './ControlChild';

export default class ControlParent extends Control {
    _template: TemplateFunction = template;
    _children: {
        child: ControlChild;
    };
    protected _timeToMountChild: number = ControlParent.timeToMountChild;
    protected _testOption: object = {
        updated: false,
    };

    protected _beforeMount(): Promise<void> | void {
        // eslint-disable-next-line no-unused-expressions, @typescript-eslint/no-unused-expressions
        new Promise((resolve) => {
            setTimeout(() => {
                this._testOption = {
                    updated: true,
                };
                resolve();
            }, ControlParent.timeToChangeOption);
        });
    }

    static readonly timeToMountChild: 30;
    static readonly timeToChangeOption: 15;
}
