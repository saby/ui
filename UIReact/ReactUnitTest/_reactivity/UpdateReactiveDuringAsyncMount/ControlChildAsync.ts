import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_reactivity/UpdateReactiveDuringAsyncMount/ControlChildAsync';

export default class ControlChildAsync extends Control {
    _template: TemplateFunction = template;
}
