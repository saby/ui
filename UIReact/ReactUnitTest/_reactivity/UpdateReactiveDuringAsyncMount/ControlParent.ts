import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_reactivity/UpdateReactiveDuringAsyncMount/ControlParent';

export default class ControlParent extends Control {
    _template: TemplateFunction = template;
    protected _value: string;
}
