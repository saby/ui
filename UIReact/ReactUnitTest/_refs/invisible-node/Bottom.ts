import { Control, TemplateFunction } from 'UI/Base';
import * as template from 'wml!ReactUnitTest/_refs/invisible-node/Bottom';

export default class Bottom extends Control {
    protected _template: TemplateFunction = template;
}
