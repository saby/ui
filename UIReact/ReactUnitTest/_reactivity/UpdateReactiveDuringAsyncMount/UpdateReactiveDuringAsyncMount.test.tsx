/**
 * @jest-environment jsdom
 */
import { render, unmountComponentAtNode } from 'react-dom';
import { act } from 'react-dom/test-utils';
import ControlParent from './ControlParent';
import ControlChildAsync from './ControlChildAsync';
import Top from './Top';
import Bottom from './Bottom';

describe('Асинхронное изменение реактивного свойства во время асинхронного построения', () => {
    let container;

    beforeEach(() => {
        /*
        _afterMount и _afterUpdate зовутся в отдельной таске, чтобы браузер мог отрисовать кадр.
        Чтобы не делать тесты асинхронными, мы просто мокнем таймеры и сами будем управлять временем.
         */
        jest.useFakeTimers();
        container = document.createElement('div');
        document.body.appendChild(container);
        jest.spyOn(window, 'requestAnimationFrame').mockImplementation(
            setTimeout
        );
    });

    afterEach(() => {
        jest.useRealTimers();
        unmountComponentAtNode(container);
        container.remove();
        container = null;
        jest.restoreAllMocks();
    });

    it('асинхронный родитель, в нём асинхронный процесс завершается раньше _beforeMount, и раньше асинхронного ребенка', async () => {
        // region Setup
        jest.spyOn(ControlParent.prototype, '_beforeMount').mockImplementation(
            function (this: ControlParent): Promise<void> {
                setTimeout(() => {
                    this._value = 'value';
                }, 4);
                return new Promise((resolve) => {
                    setTimeout(() => {
                        resolve();
                    }, 5);
                });
            }
        );
        jest.spyOn(
            ControlChildAsync.prototype,
            '_beforeMount'
        ).mockImplementation(function (this: ControlChildAsync): Promise<void> {
            return new Promise((resolve) => {
                setTimeout(() => {
                    resolve();
                }, 5);
            });
        });
        // enregion

        act(() => {
            render(<ControlParent />, container);
        });

        expect(container).toMatchSnapshot('1. Ждём Promise ControlParent');

        await act(async () => {
            jest.advanceTimersByTime(5);
        });

        expect(container).toMatchSnapshot(
            '2. Выполнился _beforeMount ControlParent, ControlChildSync построился с новым состоянием'
        );

        await act(async () => {
            jest.advanceTimersByTime(5);
        });

        expect(container).toMatchSnapshot(
            '3. ControlChildAsync построился с актуальным состоянием. Всё отрисовалось'
        );
    });

    it('асинхронный родитель, в нём асинхронный процесс завершается позже _beforeMount, но раньше асинхронного ребенка', async () => {
        // region Setup
        jest.spyOn(ControlParent.prototype, '_beforeMount').mockImplementation(
            function (this: ControlParent): Promise<void> {
                setTimeout(() => {
                    this._value = 'value';
                }, 6);
                return new Promise((resolve) => {
                    setTimeout(() => {
                        resolve();
                    }, 5);
                });
            }
        );
        jest.spyOn(
            ControlChildAsync.prototype,
            '_beforeMount'
        ).mockImplementation(function (this: ControlChildAsync): Promise<void> {
            return new Promise((resolve) => {
                setTimeout(() => {
                    resolve();
                }, 5);
            });
        });
        // enregion

        act(() => {
            render(<ControlParent />, container);
        });

        expect(container).toMatchSnapshot('1. Ждём Promise ControlParent');

        await act(async () => {
            jest.advanceTimersByTime(5);
        });

        expect(container).toMatchSnapshot(
            '2. Всё пустое, т.к. выполнился _beforeMount ControlParent, но он еще не обновил состояние'
        );

        await act(async () => {
            jest.advanceTimersByTime(1);
        });

        expect(container).toMatchSnapshot(
            '3. Несмотря на то что у ControlParent состояние изменилось, всё пустое, т.к. ждем асинхронного ControlChildAsync'
        );

        await act(async () => {
            jest.advanceTimersByTime(4);
        });

        expect(container).toMatchSnapshot(
            '4. ControlChildAsync построился, но новое состояние ControlParent не проброшено до детей'
        );

        // т.к. в _forceUpdate delay внутри Promise необходимо ждать 2 асинхронных тика
        await act(async () => {
            jest.advanceTimersByTime(0);
        });
        await act(async () => {
            jest.advanceTimersByTime(0);
        });

        await act(async () => {
            jest.runOnlyPendingTimers();
        });

        expect(container).toMatchSnapshot(
            '5. Всё отрисовалось, т.к. произошло обновление'
        );
    });

    it('асинхронный родитель, в нём асинхронный процесс завершается позже _beforeMount родителя, и позже асинхронного ребенка', async () => {
        // region Setup
        jest.spyOn(ControlParent.prototype, '_beforeMount').mockImplementation(
            function (this: ControlParent): Promise<void> {
                setTimeout(() => {
                    this._value = 'value';
                }, 11);
                return new Promise((resolve) => {
                    setTimeout(() => {
                        resolve();
                    }, 5);
                });
            }
        );
        jest.spyOn(
            ControlChildAsync.prototype,
            '_beforeMount'
        ).mockImplementation(function (this: ControlChildAsync): Promise<void> {
            return new Promise((resolve) => {
                setTimeout(() => {
                    resolve();
                }, 5);
            });
        });
        // enregion

        act(() => {
            render(<ControlParent />, container);
        });

        expect(container).toMatchSnapshot('1. Ждём Promise ControlParent');

        await act(async () => {
            jest.advanceTimersByTime(5);
        });

        expect(container).toMatchSnapshot(
            '2. Всё пустое, т.к. выполнился _beforeMount ControlParent, но у него еще старое состояние'
        );

        await act(async () => {
            jest.advanceTimersByTime(5);
        });

        expect(container).toMatchSnapshot(
            '3. ControlChildAsync построился, но в ControlParent еще старое состояние'
        );

        await act(async () => {
            jest.advanceTimersByTime(0);
        });

        expect(container).toMatchSnapshot(
            '4. Несмотря на то что у ControlParent состояние изменилось, всё пустое, т.к. не проброшено до детей'
        );

        act(() => {
            jest.advanceTimersByTime(2);
        });

        expect(container).toMatchSnapshot(
            '5. Всё отрисовалось, т.к. произошло обновление'
        );
    });

    it('синхронный родитель, в нём асинхронный процесс завершается раньше асинхронного ребенка', async () => {
        // region Setup
        jest.spyOn(ControlParent.prototype, '_beforeMount').mockImplementation(
            function (this: ControlParent): void {
                setTimeout(() => {
                    this._value = 'value';
                }, 4);
            }
        );
        jest.spyOn(
            ControlChildAsync.prototype,
            '_beforeMount'
        ).mockImplementation(function (this: ControlChildAsync): Promise<void> {
            return new Promise((resolve) => {
                setTimeout(() => {
                    resolve();
                }, 5);
            });
        });
        // enregion

        act(() => {
            render(<ControlParent />, container);
        });

        expect(container).toMatchSnapshot(
            '1. Всё пустое, т.к. у ControlParent еще старое состояние'
        );

        await act(async () => {
            jest.advanceTimersByTime(4);
        });

        expect(container).toMatchSnapshot(
            '2. Несмотря на то что у ControlParent состояние изменилось, всё пустое, т.к. не проброшено до детей'
        );

        // дожидаемся в общей сумме 5 мс взведенного setTimeout
        await act(async () => {
            jest.advanceTimersByTime(1);
        });
        // дожидаемся последующей перерисовки
        await act(async () => {
            jest.advanceTimersByTime(1);
        });

        expect(container).toMatchSnapshot(
            '3. ControlChildAsync построился и всё отрисовалось, т.к. произошло обновление'
        );
    });

    it('У контрола в _beforeMount асинхронное изменение свойства, оно должно примениться после маунтинга', async () => {
        let _beforeUpdateCalled = false;
        jest.spyOn(Bottom.prototype, '_beforeUpdate').mockImplementation(
            function (this: Bottom): void {
                _beforeUpdateCalled = true;
            }
        );

        act(() => {
            render(<Top />, container);
        });

        // дожидаемся асинхронщины Bottom
        await act(async () => {
            jest.advanceTimersByTime(20);
        });
        // дожидаемся перерисовки в качестве обновления из-за изменения свойства _value
        await act(async () => {
            jest.advanceTimersByTime(1);
        });

        expect(_beforeUpdateCalled).toBe(true);
    });

    it('синхронный родитель, в нём асинхронный процесс завершается позже асинхронного ребенка', async () => {
        // region Setup
        jest.spyOn(ControlParent.prototype, '_beforeMount').mockImplementation(
            function (this: ControlParent): Promise<void> {
                setTimeout(() => {
                    this._value = 'value';
                }, 6);
            }
        );
        jest.spyOn(
            ControlChildAsync.prototype,
            '_beforeMount'
        ).mockImplementation(function (this: ControlChildAsync): Promise<void> {
            return new Promise((resolve) => {
                setTimeout(() => {
                    resolve();
                }, 5);
            });
        });
        // enregion

        act(() => {
            render(<ControlParent />, container);
        });

        expect(container).toMatchSnapshot(
            '1. Всё пустое, т.к. у ControlParent еще старое состояние'
        );

        await act(async () => {
            jest.advanceTimersByTime(5);
        });

        expect(container).toMatchSnapshot(
            '2. ControlChildAsync построился, но в ControlParentAsync еще старое состояние'
        );

        await act(async () => {
            jest.advanceTimersByTime(0);
        });

        expect(container).toMatchSnapshot(
            '3. Несмотря на то что у ControlParentAsync состояние изменилось, всё пустое, т.к. не проброшено до детей'
        );

        act(() => {
            jest.advanceTimersByTime(1);
        });

        expect(container).toMatchSnapshot(
            '4. Всё отрисовалось, т.к. произошло обновление'
        );
    });
});
