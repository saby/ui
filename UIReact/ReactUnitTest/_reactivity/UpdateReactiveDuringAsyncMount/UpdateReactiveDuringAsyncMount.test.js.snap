// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения асинхронный родитель, в нём асинхронный процесс завершается позже _beforeMount родителя, и позже асинхронного ребенка: 1. Ждём Promise ControlParent 1`] = `
<div>
  <div
    name="$$wasaby$async$loading"
  />
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения асинхронный родитель, в нём асинхронный процесс завершается позже _beforeMount родителя, и позже асинхронного ребенка: 2. Всё пустое, т.к. выполнился _beforeMount ControlParent, но у него еще старое состояние 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      tabindex="0"
    />
    <div
      name="$$wasaby$async$loading"
    />
  </div>
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения асинхронный родитель, в нём асинхронный процесс завершается позже _beforeMount родителя, и позже асинхронного ребенка: 3. ControlChildAsync построился, но в ControlParent еще старое состояние 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      tabindex="0"
    />
    <div
      tabindex="0"
    />
  </div>
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения асинхронный родитель, в нём асинхронный процесс завершается позже _beforeMount родителя, и позже асинхронного ребенка: 4. Несмотря на то что у ControlParent состояние изменилось, всё пустое, т.к. не проброшено до детей 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      tabindex="0"
    />
    <div
      tabindex="0"
    />
  </div>
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения асинхронный родитель, в нём асинхронный процесс завершается позже _beforeMount родителя, и позже асинхронного ребенка: 5. Всё отрисовалось, т.к. произошло обновление 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      tabindex="0"
    >
      value
    </div>
    <div
      tabindex="0"
    >
      value
    </div>
  </div>
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения асинхронный родитель, в нём асинхронный процесс завершается позже _beforeMount, но раньше асинхронного ребенка: 1. Ждём Promise ControlParent 1`] = `
<div>
  <div
    name="$$wasaby$async$loading"
  />
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения асинхронный родитель, в нём асинхронный процесс завершается позже _beforeMount, но раньше асинхронного ребенка: 2. Всё пустое, т.к. выполнился _beforeMount ControlParent, но он еще не обновил состояние 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      tabindex="0"
    />
    <div
      name="$$wasaby$async$loading"
    />
  </div>
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения асинхронный родитель, в нём асинхронный процесс завершается позже _beforeMount, но раньше асинхронного ребенка: 3. Несмотря на то что у ControlParent состояние изменилось, всё пустое, т.к. ждем асинхронного ControlChildAsync 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      tabindex="0"
    />
    <div
      name="$$wasaby$async$loading"
    />
  </div>
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения асинхронный родитель, в нём асинхронный процесс завершается позже _beforeMount, но раньше асинхронного ребенка: 4. ControlChildAsync построился, но новое состояние ControlParent не проброшено до детей 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      tabindex="0"
    />
    <div
      tabindex="0"
    />
  </div>
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения асинхронный родитель, в нём асинхронный процесс завершается позже _beforeMount, но раньше асинхронного ребенка: 5. Всё отрисовалось, т.к. произошло обновление 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      tabindex="0"
    >
      value
    </div>
    <div
      tabindex="0"
    >
      value
    </div>
  </div>
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения асинхронный родитель, в нём асинхронный процесс завершается раньше _beforeMount, и раньше асинхронного ребенка: 1. Ждём Promise ControlParent 1`] = `
<div>
  <div
    name="$$wasaby$async$loading"
  />
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения асинхронный родитель, в нём асинхронный процесс завершается раньше _beforeMount, и раньше асинхронного ребенка: 2. Выполнился _beforeMount ControlParent, ControlChildSync построился с новым состоянием 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      tabindex="0"
    >
      value
    </div>
    <div
      name="$$wasaby$async$loading"
    />
  </div>
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения асинхронный родитель, в нём асинхронный процесс завершается раньше _beforeMount, и раньше асинхронного ребенка: 3. ControlChildAsync построился с актуальным состоянием. Всё отрисовалось 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      tabindex="0"
    >
      value
    </div>
    <div
      tabindex="0"
    >
      value
    </div>
  </div>
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения синхронный родитель, в нём асинхронный процесс завершается позже асинхронного ребенка: 1. Всё пустое, т.к. у ControlParent еще старое состояние 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      tabindex="0"
    />
    <div
      name="$$wasaby$async$loading"
    />
  </div>
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения синхронный родитель, в нём асинхронный процесс завершается позже асинхронного ребенка: 2. ControlChildAsync построился, но в ControlParentAsync еще старое состояние 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      tabindex="0"
    />
    <div
      tabindex="0"
    />
  </div>
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения синхронный родитель, в нём асинхронный процесс завершается позже асинхронного ребенка: 3. Несмотря на то что у ControlParentAsync состояние изменилось, всё пустое, т.к. не проброшено до детей 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      tabindex="0"
    />
    <div
      tabindex="0"
    />
  </div>
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения синхронный родитель, в нём асинхронный процесс завершается позже асинхронного ребенка: 4. Всё отрисовалось, т.к. произошло обновление 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      tabindex="0"
    >
      value
    </div>
    <div
      tabindex="0"
    >
      value
    </div>
  </div>
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения синхронный родитель, в нём асинхронный процесс завершается раньше асинхронного ребенка: 1. Всё пустое, т.к. у ControlParent еще старое состояние 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      tabindex="0"
    />
    <div
      name="$$wasaby$async$loading"
    />
  </div>
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения синхронный родитель, в нём асинхронный процесс завершается раньше асинхронного ребенка: 2. Несмотря на то что у ControlParent состояние изменилось, всё пустое, т.к. не проброшено до детей 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      tabindex="0"
    />
    <div
      name="$$wasaby$async$loading"
    />
  </div>
</div>
`;

exports[`Асинхронное изменение реактивного свойства во время асинхронного построения синхронный родитель, в нём асинхронный процесс завершается раньше асинхронного ребенка: 3. ControlChildAsync построился и всё отрисовалось, т.к. произошло обновление 1`] = `
<div>
  <div
    tabindex="0"
  >
    <div
      tabindex="0"
    >
      value
    </div>
    <div
      tabindex="0"
    >
      value
    </div>
  </div>
</div>
`;
